### PR TITLE
[CARBONDATA-393] implement test cases for core.keygenerator module 

### PR DESCRIPTION
**Pull request by anuragknoldus**
In this PR, I have implemented unit test cases for core.keygenerator package.

### CODE REVIEW

File: core/src/test/java/org/apache/carbondata/core/keygenerator/columnar/impl/MultiDimKeyVarLengthEquiSplitGeneratorUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.Arrays;
+
+public class MultiDimKeyVarLengthEquiSplitGeneratorUnitTest {
+
+    private MultiDimKeyVarLengthEquiSplitGenerator multiDimKeyVarLengthEquiSplitGenerator;
+
+    @Before
+    public void setup() {
+        int[] lens = new int[]{1, 2, 3, 4, 5, 7, 8, 9, 0, 9, 8, 7, 6, 5, 4, 3};
+        byte b = 100;
```
Comment (ravipesala) [inline core/src/test/java/org/apache/carbondata/core/keygenerator/columnar/impl/MultiDimKeyVarLengthEquiSplitGeneratorUnitTest.java]: Add more testcase with `dimensionsToSplit` as 1, we mostly us this scenerio.

File: core/src/test/java/org/apache/carbondata/core/keygenerator/columnar/impl/MultiDimKeyVarLengthVariableSplitGeneratorUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.Arrays;
+
+public class MultiDimKeyVarLengthVariableSplitGeneratorUnitTest {
+
+    private MultiDimKeyVarLengthVariableSplitGenerator multiDimKeyVarLengthVariableSplitGenerator;
+
+    @Before
+    public void setup() {
+        int[] lens = new int[]{1, 2, 3, 4, 5, 7, 8, 9, 0, 9, 8, 7, 6, 5, 4, 3};
+        int[] dimSplit = new int[]{50, 30};
```
Comment (ravipesala) [inline core/src/test/java/org/apache/carbondata/core/keygenerator/columnar/impl/MultiDimKeyVarLengthVariableSplitGeneratorUnitTest.java]: Here we should give proper `dimSplit`  and add more testcases

File: core/src/test/java/org/apache/carbondata/core/keygenerator/mdkey/BitsUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+
+public class BitsUnitTest {
+    private Bits bits;
+
+    @Test
+    public void testGetKeyByteOffsets() throws Exception {
+        int[] lens = new int[]{1, 2, 3};
```
Comment (ravipesala) [inline core/src/test/java/org/apache/carbondata/core/keygenerator/mdkey/BitsUnitTest.java]: Add more testcases with big values and also cover the boundary conditions in test cases.

File: core/src/test/java/org/apache/carbondata/core/keygenerator/mdkey/BitsUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        int[] expected = new int[]{0, 0};
+        int[] result = bits.getKeyByteOffsets(index);
+        assertThat(result, is(equalTo(expected)));
+    }
+
+    @Test
+    public void testGetWithIntKeys() throws Exception {
+        int[] lens = new int[]{20, 35, 10};
+        bits = new Bits(lens);
+        long[] expected = new long[]{703687441812490L, 0};
```
Comment (ravipesala) [inline core/src/test/java/org/apache/carbondata/core/keygenerator/mdkey/BitsUnitTest.java]: test with negative and boundary conditions

**Comment by ravipesala** [comment]
Please format the classes

**Comment by anuragknoldus** [comment]
Hi @ravipesala,
I have made changes which proposed. Could you please review it ?

**Comment by chenliang613** [comment]
@anuragknoldus the below info only be template for reminding to check everything, the PR description doesn't need to include , please remove it.

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

Make sure the PR title is formatted like:
[CARBONDATA-] Description of pull request

Make sure tests pass via mvn clean verify. (Even better, enable
Travis-CI on your fork and ensure the whole test matrix passes).

Replace in the title with the actual Jira issue
number, if there is one.

If this contribution is large, please file an Apache
Individual Contributor License Agreement.

Testing done

Please provide details on

Whether new unit test cases have been added or why no new tests are required?
What manual testing you have done?
Any additional information to help reviewers in testing this change. For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

**Comment by anuragknoldus** [comment]
OK @chenliang613, I have unchecked all boxes in the template.

File: core/src/test/java/org/apache/carbondata/core/keygenerator/columnar/impl/MultiDimKeyVarLengthEquiSplitGeneratorUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+import java.util.Arrays;
+
+public class MultiDimKeyVarLengthEquiSplitGeneratorUnitTest {
+
+  static MultiDimKeyVarLengthEquiSplitGenerator multiDimKeyVarLengthEquiSplitGenerator;
+
+  @BeforeClass public static void setup() {
+    int[] lens = new int[] { 32, 8, 16, 16, 16 };
+    byte dimensionsToSplit = 1;
```
Comment (ravipesala) [inline core/src/test/java/org/apache/carbondata/core/keygenerator/columnar/impl/MultiDimKeyVarLengthEquiSplitGeneratorUnitTest.java]: And we should test actual logic by using method `generateAndSplitKey`  and `getKeyArray`

File: core/src/test/java/org/apache/carbondata/core/keygenerator/columnar/impl/MultiDimKeyVarLengthVariableSplitGeneratorUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.core.Is.is;
+
+public class MultiDimKeyVarLengthVariableSplitGeneratorUnitTest {
+
+  static MultiDimKeyVarLengthVariableSplitGenerator multiDimKeyVarLengthVariableSplitGenerator;
+
+  @BeforeClass public static void setup() {
+    int[] lens = new int[] { 32, 8, 16, 16, 16 };
+    int[] dimSplit = new int[] { 1, 1, 1, 1, 1 };
```
Comment (ravipesala) [inline core/src/test/java/org/apache/carbondata/core/keygenerator/columnar/impl/MultiDimKeyVarLengthVariableSplitGeneratorUnitTest.java]: Please add testcases for different `dimSplit`

Comment (ravipesala) [inline core/src/test/java/org/apache/carbondata/core/keygenerator/columnar/impl/MultiDimKeyVarLengthVariableSplitGeneratorUnitTest.java]: And add testcases for boundary and negative flows like what if number oberflows the bit range

**Comment by ravipesala** [comment]
Can one of the admins verify this patch?

**Comment by ravipesala** [comment]
add to whitelist

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.5.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/77/

**Comment by ravipesala** [comment]
LGTM
